network/transport/grpc/addressbook.go

```diff
@@ -0,0 +1,226 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package grpc
+
+import (
+	"errors"
+	"fmt"
+	"github.com/nuts-foundation/nuts-node/core"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/nuts-foundation/go-stoabs"
+	"github.com/nuts-foundation/nuts-node/network/transport"
+)
+
+// AddressBook provides an API for protocols to query the ConnectionManager's known addresses.
+type AddressBook interface {
+	// Update the address for the peer's existing getContact, or create one if it does not exist.
+	// If the address is empty, the contact is removed.
+	// bootstrap nodes contain an address with an empty DID.
+	Update(peer transport.Peer) error
+	// Get the getContact if it exists, or returns false if it does not.
+	Get(peer transport.Peer) (*contact, bool)
+	// All returns a copy of the contacts of connectors.
+	All() []*contact
+}
+
+func newAddressBook(connectionStore stoabs.KVStore, backoffCreator func() Backoff, hasNoConnection predicate) *addressBook {
+	return &addressBook{
+		contacts:        make([]*contact, 0),
+		backoffStore:    connectionStore,
+		backoffCreator:  backoffCreator,
+		hasNoConnection: hasNoConnection,
+	}
+}
+
+type addressBook struct {
+	mux             sync.RWMutex // TODO: check race conditions
+	contacts        []*contact
+	backoffStore    stoabs.KVStore
+	backoffCreator  func() Backoff
+	hasNoConnection predicate
+}
+
+func (a *addressBook) Get(peer transport.Peer) (*contact, bool) {
+	a.mux.RLock()
+	defer a.mux.RUnlock()
+	return a.get(peer)
+}
+
+func (a *addressBook) get(peer transport.Peer) (*contact, bool) {
+	if !peer.NodeDID.Empty() { // find on DID
+		for _, o := range a.contacts {
+			if peer.NodeDID.Equals(o.peer.NodeDID) {
+				return o, true
+			}
+		}
+	} else { // find on address -> bootstrap only
+		for _, o := range a.contacts {
+			if peer.Address == o.peer.Address {
+				return o, true
+			}
+		}
+	}
+	return nil, false
+}
+
+func (a *addressBook) Update(peer transport.Peer) error {
+	a.mux.Lock()
+	defer a.mux.Unlock()
+
+	// TODO: should this check include url validation?
+	if peer.NodeDID.Empty() && peer.Address == "" {
+		return errors.New("invalid peer")
+	}
+
+	// update existing address
+	current, exists := a.get(peer)
+	if exists {
+		if peer.Address == current.peer.Address {
+			// nothing to update
+		} else if peer.Address == "" {
+			// delete NutsComm -> delete dialer
+			a.remove(current)
+		} else {
+			// update DID's address and reset backoff
+			current.peer.Address = peer.Address
+			current.backoff.Reset(0)
+		}
+		return nil
+	}
+
+	// add new address
+	backoff := a.backoffCreator()
+	if !peer.NodeDID.Empty() {
+		// only persist non-bootstrap.
+		// store the backoff under the DID since an address could be used by multiple DIDs.
+		backoff = NewPersistedBackoff(a.backoffStore, fmt.Sprintf("did:%s:%s", peer.NodeDID.Method, peer.NodeDID.ID), backoff)
```

Comment:
because `String()` may contain a path or fragment. The DID will be taken from the document though, so it should be clean. will change to use the `String()`